showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java

```diff
@@ -0,0 +1,208 @@
+        ApiMethodDescriptor<RequestT, ResponseT> method,
+        HttpJsonCallOptions callOptions,
+        HttpJsonChannel next) {
+      HttpJsonClientCall<RequestT, ResponseT> call = next.newCall(method, callOptions);
+      return new ForwardingHttpJsonClientCall.SimpleForwardingHttpJsonClientCall<
+          RequestT, ResponseT>(call) {
+        @Override
+        public void start(Listener<ResponseT> responseListener, HttpJsonMetadata requestHeaders) {
+          Listener<ResponseT> forwardingResponseListener =
+              new ForwardingHttpJsonClientCallListener.SimpleForwardingHttpJsonClientCallListener<
+                  ResponseT>(responseListener) {
+                @Override
+                public void onHeaders(HttpJsonMetadata responseHeaders) {
+                  metadata = responseHeaders;
+                  super.onHeaders(responseHeaders);
+                }
+
+                @Override
+                public void onMessage(ResponseT message) {
+                  super.onMessage(message);
+                }
+
+                @Override
+                public void onClose(int statusCode, HttpJsonMetadata trailers) {
+                  super.onClose(statusCode, trailers);
+                }
+              };
+
+          super.start(forwardingResponseListener, requestHeaders);
+        }
+      };
+    }
+  }
+
+  private HttpJsonCapturingClientInterceptor httpJsonInterceptor;
+  private GrpcCapturingClientInterceptor grpcInterceptor;
+  private HttpJsonCapturingClientInterceptor httpJsonComplianceInterceptor;
+  private GrpcCapturingClientInterceptor grpcComplianceInterceptor;
+  private EchoClient grpcClient;
+  private EchoClient httpJsonClient;
+  private ComplianceClient grpcComplianceClient;
+  private ComplianceClient httpJsonComplianceClient;
+
+  @Before
+  public void createClients() throws Exception {
+    // Create gRPC Interceptor and Client
+    grpcInterceptor = new GrpcCapturingClientInterceptor();
+    grpcClient = TestClientInitializer.createGrpcEchoClient(ImmutableList.of(grpcInterceptor));
+
+    // Create HttpJson Interceptor and Client
+    httpJsonInterceptor = new HttpJsonCapturingClientInterceptor();
+    httpJsonClient =
+        TestClientInitializer.createHttpJsonEchoClient(ImmutableList.of(httpJsonInterceptor));
+
+    // Create gRPC ComplianceClient and Interceptor
+    // Creating a compliance client to test case where api version is not set
+    grpcComplianceInterceptor = new GrpcCapturingClientInterceptor();
+    grpcComplianceClient =
+        TestClientInitializer.createGrpcComplianceClient(
+            ImmutableList.of(grpcComplianceInterceptor));
+
+    // Create HttpJson ComplianceClient and Interceptor
+    httpJsonComplianceInterceptor = new HttpJsonCapturingClientInterceptor();
+    httpJsonComplianceClient =
+        TestClientInitializer.createHttpJsonComplianceClient(
+            ImmutableList.of(httpJsonComplianceInterceptor));
```

Comment:
In current setup, we _could_ override with `.setHeaderProvider()` when creating the test client. I am reluctant to do so though. 
- I feel like showcase integration testing should test expected behavior of un-modified service behaviors E2E.
- As we discussed in doc comment last time, I am inclined to not allow user override this "x-goog-api-version" header with client libraries. (Will raise a separate pr) In this case, "unsets the showcase apiVersion value" should not work.

Reply:
> I feel like showcase integration testing should test expected behavior of un-modified service behaviors E2E.

IMO, it shouldn't just be limited to un-modified service behaviors. We should try to account for customizations where we can (i.e. If new features are added or showcase has new functionality). We're not in state where every modification/ feature is covered, but I think we should strive to get as close as possible. There are going to be certain spots we miss and new edge cases that are discovered, but we should try to backfill/ add them in as we discover them.

> I am inclined to not allow user override this "x-goog-api-version" header with client libraries.

Makes sense. Let's get some insight from the core team to see if this is prioritized (and if not, we can always prioritize this for Java clients ourselves). The public setter method is pubic as of now and removing will be challenging/ impossible. We can explore some options to see how we want to proceed.

Being that removing may be challenging/impossible, I think it may make sense to include both Echo and Compliance clients in then since we test for the behavior of add custom headers to both clients that that have an auto generated ApiVersion vs clients that don't have this generated.

i.e. I believe Echo with custom headers set should have a conflict error message and Compliance with custom headers should not have a conflict error message.

Reply:
I agree with most of these, but wanted to clarify a few:
- This [ITApiVersionHeaders.java](https://github.com/googleapis/sdk-platform-java/pull/2737/files/219c8fdb150a33a9f2e44c442f4e91a586000ead#diff-0c74baec42410162926db3ec4be8e8f2c04fd9d6cf214ef5bbf407e7e350f0b5) tests for behavior when client has an auto generated ApiVersion (Echo) and gets correctly send via header on requests, and when client don't have this generated (Compliance) then no such header sent. User header are not covered.
- To test behavior of adding user header to both clients also makes sense, I will add based on current behavior. This may change in the future. In fact, I have a todo item in https://github.com/googleapis/sdk-platform-java/pull/2745 for it if we decide to go with this approach.
- "The public setter method" >> I assume you are referring to [this setter](https://github.com/googleapis/sdk-platform-java/blob/c9d8ded1a8af3d7cbb9a647cb802a71412c8d7d7/gax-java/gax/src/main/java/com/google/api/gax/rpc/ApiClientHeaderProvider.java#L218)? If so, this is only one HeaderProvider the client library code uses, users could use other HeaderProvider. But to guard users to not override  "x-goog-api-version" header with `setHeaderProvider()`, I am thinking adding to conflict resolution as in [pr2745](https://github.com/googleapis/sdk-platform-java/pull/2745) 
- "Let's get some insight from the core team if this is prioritized " >> Other than comment on proposal doc, I've started a thread on the aip [here](https://github.com/aip-dev/google.aip.dev/pull/1331/files#r1591764838) 

To summarize, I am keeping both clients in this test to cover scenario of service opted-in for api-version and not. Also adding tests for setting custom api version headers for these clients, these behavior may change via a separate pr.

Reply:
> This [ITApiVersionHeaders.java](https://github.com/googleapis/sdk-platform-java/pull/2737/files/219c8fdb150a33a9f2e44c442f4e91a586000ead#diff-0c74baec42410162926db3ec4be8e8f2c04fd9d6cf214ef5bbf407e7e350f0b5) tests for behavior when client has an auto generated ApiVersion (Echo) and gets correctly send via header on requests, and when client don't have this generated (Compliance) then no such header sent. User header are not covered.

Yep, I think that's good! The test cases do look good. I don't mean that we should try to cover modifying *every* user header, but rather the relevant one to the test (api-version header).

> In fact, I have a todo item in https://github.com/googleapis/sdk-platform-java/pull/2745

Let's have an internal discussion. I do prefer this behavior but I believe this does technically put us at a "behavior breaking change" which is something that we try not to do (if possible). I think this may be a case where we can argue for it given that it's a relatively new feature.

> Also adding tests for setting custom api version headers for these clients, these behavior may change via a separate pr.

I'll do another round of review on this PR. No worries about adding modifications to the header in this PR if you haven't already. I believe it's already a pretty niche case and it can be added and/or tested based on what happens to #2745